hera_cal/frf.py

```diff
@@ -97,6 +106,312 @@ def sky_frates(uvd, keys=None, frate_standoff=0.0, frate_width_multiplier=1.0, m
+    if dfr is None:
+        # if no dfr provided, set to 1 / (ntimes * dt)
+        dfr = 1. / (dt * nfr)
+
+    # build grid.
+    fr_grid = np.arange(-nfr // 2, nfr // 2) * dfr
+
+    # frequency tapering function expected for power spectra.
+    # square b/c for power spectrum.
+    ftaper = dspec.gen_window(taper, uvd.Nfreqs) ** 2.
+    if keys is None:
+        keys = uvd.get_antpairpols()
+    unique_pols = set({})
+    for bl in keys:
+        if bl[-1] not in unique_pols:
+            unique_pols.add(bl[-1])
+
+    profiles = {}  # store profiles here.
+    ap_blkeys = {}
+    # keeps track of different polarizations for each antenna pair
+    # for if we are going to sum over polarizations.
+    # get redundancies (will only compute fr-profile once for each red group).
+    antpos, antnums = uvd.get_ENU_antpos()
+    antpos = {an: ap for an, ap in zip(antnums, antpos)}
+    reds = redcal.get_reds(antpos, pols=list(unique_pols), include_autos=True)
+    reds = [[bl for bl in rg if bl in keys or utils.reverse_bl(bl) in keys] for rg in reds]
+    reds = [rg for rg in reds if len(rg) > 0]
+
+    for redgrp in reds:
+        # only explicitly calculate fr profile for the first vis in each redgroup.
+        bl = redgrp[0]
+        echo("Generating FR-Profile of {} at {}".format(bl, str(datetime.datetime.now())), verbose=verbose)
+        # sum beams from all frequencies
+        # get polarization number
+        polnum = np.where(uvutils.polstr2num(bl[-1], x_orientation=uvb.x_orientation) == uvb.polarization_array)[0][0]
+        # get baseline vector in equitorial coordinates.
+        ind1 = np.where(antnums == bl[0])[0][0]
+        ind2 = np.where(antnums == bl[1])[0][0]
+        blvec = uvd.antenna_positions[ind2] - uvd.antenna_positions[ind1]
+        # initialize binned power.
+        # we will bin frate power together for all frequencies, weighted by taper.
+        binned_power = np.zeros_like(fr_grid)
+        binned_power_conj = np.zeros_like(fr_grid)
+        # iterate over each frequency and ftaper weighting.
+        # use linspace to make sure we get first and last frequencies.
+        chans_to_use = np.linspace(0, uvd.Nfreqs - 1, int(uvd.Nfreqs / fr_freq_skip)).astype(int)
+        for f0, fw in zip(uvd.freq_array[0, chans_to_use], ftaper[chans_to_use]):
+            frates = np.dot(np.cross(np.array([0, 0, 1.]), blvec), eq_xyz) * 2 * np.pi * f0 / SPEED_OF_LIGHT / SDAY_KSEC
+            # square of power beam values in directions of sky pixels
+            bsq = np.abs(uvb.data_array[0, 0, polnum, np.argmin(np.abs(f0 - uvb.freq_array[0])), :].squeeze()) ** 2.
+            # set beam below horizon to be zero.
+            bsq[polar >= np.pi / 2.] = 0.
+            # get fringe-rate bin membership for each pixel.
+            fr_bins = np.round(frates / dfr + nfr / 2).astype(int)
+            fr_bins_conj = np.round(frates / -dfr + nfr / 2).astype(int)
+            # bin power.
+            for binnum in range(nfr):
```

Comment:
I imagine that a lot of the reason this is slow is because of this triple-nested for loop. Can this be vectorized?

Reply:
You should be able to vectorize this with `scipy.stats.binned_statistic`. For example, I *think* this should work:
```
from scipy.stats import binned_statistic
binned_power = binned_statistic(frates, bsq * fw, statistic="sum", bins=fr_grid)[0]
binned_power_conj = binned_statistic(-frates, bsq * fw, statistic="sum", bins=fr_grid)[0]
```
If you do this (which I think is the right thing to do), then you need to make sure that you define the fringe-rate grid correctly. In particular, it specifies the *edges* of the bins, and so should have length `nfr + 1`.